### PR TITLE
[Tune] Add example and tutorial for DCGAN

### PR DESCRIPTION
**Pull request by hhbyyh**
Add the tutorial and examples
## Related issue number
https://github.com/ray-project/ray/issues/5660

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Comment by hhbyyh** [comment]
Moving from https://github.com/ray-project/ray/pull/6389 to here. Sorry for the inconvenience.

File: doc/source/tune-advanced-tutorial.rst [hunk 1]
```diff
@@ -0,0 +1,259 @@
+Tune Advanced tutorials
```
Comment (richardliaw) [inline doc/source/tune-advanced-tutorial.rst]:
```suggestion
Tune Advanced Tutorials
```

File: doc/source/tune-advanced-tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+=======================
+
+In this page, we will explore some advanced functions in Tune with more examples.
```
Comment (richardliaw) [inline doc/source/tune-advanced-tutorial.rst]:
```suggestion
In this page, we will explore some advanced functionality in Tune with more examples.
```

File: doc/source/tune-advanced-tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+.. contents::
+    :local:
+    :backlinks: none
+
+A native example of Trainable
+-----------------------------
+As mentioned in `Tune User Guide <tune-usage.html#Tune Training API>`_, Training can be done
+with either the `Trainable <tune-usage.html#trainable-api>`__ **Class API** or
+**function-based API**. Comparably, ``Trainable`` supports stateful Actor and checkpoint/restore,
```
Comment (richardliaw) [inline doc/source/tune-advanced-tutorial.rst]:
```suggestion
**function-based API**. Comparably, ``Trainable`` is stateful, supports checkpoint/restore functionality,
```

File: doc/source/tune-advanced-tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+        },
+        num_samples=10,
+        config={
+            "guess": tune.randint(1, 10000)
+        })
+
+    print('best config: ', analysis.get_best_config(metric="diff", mode="min"))
+
+The program randomly picks 10 number from [1, 10000) and finds which is closer to the password.
+As a subclass of ``ray.tune.Trainable``, Tune will convert ``Example`` into a Ray actor, which
```
Comment (richardliaw) [inline doc/source/tune-advanced-tutorial.rst]:
```suggestion
As a subclass of ``ray.tune.Trainable``, Tune will convert ``Guesser`` into a Ray actor, which
```

File: doc/source/tune-advanced-tutorial.rst [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+   :end-before: __trainable_end__
+
+The example reuses some of the functions in ray/tune/examples/mnist_pytorch.py, and is also a good
+demo for how to decouple the tuning logic and original training code.
+
+Here, we also override ``reset_config``. This method is optional but can be implemented to speed
+up algorithms such as PBT, and to allow performance optimizations such as running experiments
+with ``reuse_actors=True``.
+
+Then we define a PBT scheduler
```
Comment (richardliaw) [inline doc/source/tune-advanced-tutorial.rst]:
```suggestion
Then, we define a PBT scheduler:
```

File: doc/source/tune-advanced-tutorial.rst [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+
+.. image:: images/tune_advanced_plot1.png
+
+DCGAN with Trainable and PBT
+----------------------------
+
+The Generative Adversarial Networks (GAN) (Goodfellow et al., 2014) framework learns generative
+models via a training paradigm consisting of two competing modules – a generator and a
+discriminator. GAN training can be remarkably brittle and unstable in the face of suboptimal
+hyper- parameter selection with generators often collapsing to a single mode or diverging entirely.
```
Comment (richardliaw) [inline doc/source/tune-advanced-tutorial.rst]:
```suggestion
hyperparameter selection with generators often collapsing to a single mode or diverging entirely.
```

File: doc/source/tune-advanced-tutorial.rst [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+As presented in `Population Based Training (PBT) <https://deepmind.com/blog/population-based-training-neural-networks>`__,
+PBT can help with the DCGAN training and next we're gonna to build a simple DCGAN demo.
```
Comment (richardliaw) [inline doc/source/tune-advanced-tutorial.rst]:
```suggestion
PBT can help with the DCGAN training. We will now walkthrough how to do this in Tune.
```

File: python/ray/tune/examples/pbt_convnet_example.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--smoke-test", action="store_true", help="Finish quickly for testing")
+    args, _ = parser.parse_known_args()
+
+    ray.init()
+    datasets.MNIST("~/data", train=True, download=True)
+
+    validate_save_restore(PytorchTrainble)
```
Comment (richardliaw) [inline python/ray/tune/examples/pbt_convnet_example.py]: Can you add a comment here?

File: python/ray/tune/examples/pbt_dcgan_mnist/pbt_dcgan_mnist.py [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+
+# __INCEPTION_SCORE_end__
+
+
+def train(netD, netG, optimG, optimD, criterion, dataloader, iteration):
+    real_label = 1
+    fake_label = 0
+
+    for i, data in enumerate(dataloader, 0):
+        if i >= 5:
```
Comment (richardliaw) [inline python/ray/tune/examples/pbt_dcgan_mnist/pbt_dcgan_mnist.py]: Can you make this a global variable?

File: python/ray/tune/examples/pbt_dcgan_mnist/pbt_dcgan_mnist.py [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+            break
+
+        netD.zero_grad()
+        real_cpu = data[0].to(device)
+        b_size = real_cpu.size(0)
+        label = torch.full((b_size, ), real_label, device=device)
```
Comment (richardliaw) [inline python/ray/tune/examples/pbt_dcgan_mnist/pbt_dcgan_mnist.py]: I think device should be passed in.

Comment (richardliaw) [reply python/ray/tune/examples/pbt_dcgan_mnist/pbt_dcgan_mnist.py]: `torch.device()` should be the object used. That will make it easier to interface with GPU.

**Review by richardliaw — COMMENTED** [review]
Left my last comments. Did you check if this works on GPU?

File: python/ray/tune/examples/pbt_convnet_example.py (hunk 8 above)
Comment (hhbyyh) [reply python/ray/tune/examples/pbt_convnet_example.py]: done

File: python/ray/tune/examples/pbt_dcgan_mnist/pbt_dcgan_mnist.py (hunk 10 above)
Comment (hhbyyh) [reply python/ray/tune/examples/pbt_dcgan_mnist/pbt_dcgan_mnist.py]: updated.

**Comment by hhbyyh** [comment]
Updated. Thanks for the comments. I don't have a GPU env right now. Maybe sometime next month...

**Review by richardliaw — APPROVED** [review]

**Comment by hhbyyh** [comment]
Hi @richardliaw From the CI log
```
File "/opt/conda/lib/python3.6/os.py", line 220, in makedirs
    mkdir(name, mode)
FileExistsError: [Errno 17] File exists: '/root/ray_results'
FAILED 1
```
Is it a synchronized issue? i.e. multiple process checking and creating folder at the same time.

File: doc/source/tune-advanced-tutorial.rst (hunk 2 above)
Comment (hhbyyh) [reply doc/source/tune-advanced-tutorial.rst]: updated.

File: doc/source/tune-advanced-tutorial.rst (hunk 3 above)
Comment (hhbyyh) [reply doc/source/tune-advanced-tutorial.rst]: updated.

File: doc/source/tune-advanced-tutorial.rst (hunk 4 above)
Comment (hhbyyh) [reply doc/source/tune-advanced-tutorial.rst]: Updated.

**Comment by hhbyyh** [comment]
I sent a PR try to fix the race condition when creating the log folder. https://github.com/ray-project/ray/pull/6420.

**Comment by richardliaw** [comment]
Merged it! Can you try rebase?

**Comment by hhbyyh** [comment]
Thanks. Rebased this branch.

**Comment by richardliaw** [comment]
jenkins test tune

**Comment by hhbyyh** [comment]
I checked the error from CI but seems it's an unrelated setup issue. Let me know if it's related to this PR.. Thanks.

**Comment by richardliaw** [comment]
jenkins retest this please

**Comment by richardliaw** [comment]
```
Traceback (most recent call last):
  File "/ray/python/ray/tune/examples/pbt_dcgan_mnist/pbt_dcgan_mnist.py", line 318, in <module>
    mnist_cnn.load_state_dict(torch.load("./mnist_cnn.pt"))
  File "/opt/conda/lib/python3.6/site-packages/torch/serialization.py", line 419, in load
    f = open(f, 'rb')
FileNotFoundError: [Errno 2] No such file or directory: './mnist_cnn.pt'
```

This should instead be:
```
    path = os.path.dirname(ray.__file__)
    model_path = os.path.join(
        path, "tune/examples/pbt_dcgan_mnist/mnist_cnn.pt")
    mnist_cnn.load_state_dict(torch.load(model_path))
```

**Comment by hhbyyh** [comment]
Thanks. Updated.

**Comment by richardliaw** [comment]
jenkins test tune

**Comment by richardliaw** [comment]
Looks good to me! Thanks for the contribution!

**Comment by hhbyyh** [comment]
Thanks for helping with review!

**Pull request merged**